OMCompiler/SimulationRuntime/c/simulation/solver/jacobianSymbolical.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * This file is part of OpenModelica.
+ *
+ * Copyright (c) 1998-2016, Open Source Modelica Consortium (OSMC),
+ * c/o Linköpings universitet, Department of Computer and Information Science,
+ * SE-58183 Linköping, Sweden.
+ *
+ * All rights reserved.
+ *
+ * THIS PROGRAM IS PROVIDED UNDER THE TERMS OF THE BSD NEW LICENSE OR THE
+ * GPL VERSION 3 LICENSE OR THE OSMC PUBLIC LICENSE (OSMC-PL) VERSION 1.2.
+ * ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS PROGRAM CONSTITUTES
+ * RECIPIENT'S ACCEPTANCE OF THE OSMC PUBLIC LICENSE OR THE GPL VERSION 3,
+ * ACCORDING TO RECIPIENTS CHOICE.
+ *
+ * The OpenModelica software and the OSMC (Open Source Modelica Consortium)
+ * Public License (OSMC-PL) are obtained from OSMC, either from the above
+ * address, from the URLs: http://www.openmodelica.org or
+ * http://www.ida.liu.se/projects/OpenModelica, and in the OpenModelica
+ * distribution. GNU version 3 is obtained from:
+ * http://www.gnu.org/copyleft/gpl.html. The New BSD License is obtained from:
+ * http://www.opensource.org/licenses/BSD-3-Clause.
+ *
+ * This program is distributed WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, EXCEPT AS
+ * EXPRESSLY SET FORTH IN THE BY RECIPIENT SELECTED SUBSIDIARY LICENSE
+ * CONDITIONS OF OSMC-PL.
+ *
+ */
+
+ /*! \file jacobian_symbolical.c
+ */
+
+#include "simulation/solver/jacobianSymbolical.h"
+
+
+void genericParallelColoredSymbolicJacobianEvaluation(int rows, int columns, SPARSE_PATTERN* spp,
+                                                      void* matrixA, ANALYTIC_JACOBIAN* jacColumns, DATA* data,
+                                                      threadData_t* threadData,
+                                                      void (*f)(int, int, int, double, void*, int))
+{
+  ANALYTIC_JACOBIAN* t_jac = jacColumns;
+
+  unsigned int ii, j, l, nth, i;
+  for(unsigned int i = 0; i < spp->maxColors; i++) {
+    for(unsigned int j=0; j < columns; j++) {
+      if(spp->colorCols[j]-1 == i)
+        t_jac->seedVars[j] = 1;
+    }
+    data->callback->functionJacA_column(data, threadData, t_jac, NULL);
+    for(unsigned int j = 0; j < columns; j++) {
+      if(t_jac->seedVars[j] == 1) {
+        nth = spp->leadindex[j];
+        while(nth < spp->leadindex[j+1]) {
+          l = spp->index[nth];
+          (*f)(l, j, nth, t_jac->resultVars[l], matrixA, rows);
+          nth++;
+        };
+      }
+    }
+
+    for(j=0; j < columns; j++)
```

Comment:
Uses j at the toplevel, but you might very well write unsigned j = 0 here directly to simply the top decl

Reply:
I'm more a fan of writing all variables at the top instead of in each scoope.

Reply:
> 
> 
> I'm more a fan of writing all variables at the top instead of in each scoope.

I buy that completely, everyone has his or her own style! However, try to keep it in one way over the other. Currently, it is a bit asymmetrical that you use  the top declaration in the last loop but declare variables with the same name in the scope preceding it

Reply:
Well I did change it but for whatever reason it was not showing on this PR. But now it should be better.

Reply:
I have tried quite a lot to not declare variables in the for-loop in our C-sources. It's not allowed in C89:
```
a.c:3:9: error: variable declaration in for loop is a C99-specific feature
      [-Werror,-Wc99-extensions]
   for (int i=0; i<3; i++) {
        ^
1 error generated.
```
It's especially important for source-code FMUs as some older C-compilers for embedded systems doesn't like it (and some Visual Studio versions as well as I recall).